lhotse/features.py

```diff
@@ -365,59 +364,113 @@ def _process_and_store_recording(
-        #    the signals to satisfy requested SNR
-        left_energy, right_energy = [np.sum(np.exp(f)) for f in [left_feats, right_feats]]
-        requested_right_energy = left_energy * (10.0 ** (-snr / 10))
-        overlayed_feats = np.log(
-            np.exp(padded_left_feats) +
-            (requested_right_energy / right_energy) * np.exp(padded_right_feats)
+            offset,
+            frame_length=self.frame_length,
+            frame_shift=self.frame_shift
         )
-    else:
-        overlayed_feats = np.log(np.exp(padded_left_feats) + np.exp(padded_right_feats))
-
-    return overlayed_feats
+        current_num_frames = self.mixed_feats.shape[0]
+        incoming_num_frames = feats.shape[0] + num_frames_offset
+        mix_num_frames = max(current_num_frames, incoming_num_frames)
+
+        existing_feats = self.mixed_feats
+        feats_to_add = feats
+
+        # When the existing frames are less than what we anticipate after the mix,
+        # we need to pad after the end of the existing features mixed so far.
+        if current_num_frames < mix_num_frames:
+            existing_feats = np.vstack([
+                self.mixed_feats,
+                self.log_energy_floor * np.ones((mix_num_frames - current_num_frames, self.num_features))
+            ])
+
+        # When there is an offset, we need to pad before the start of the features we're adding.
+        if offset > 0:
+            feats_to_add = np.vstack([
+                self.log_energy_floor * np.ones((num_frames_offset, self.num_features)),
+                feats_to_add
+            ])
+
+        # When the features we're mixing in are shorter that the anticipated mix length,
+        # we need to pad after their end.
+        # Note: we're doing that non-efficiently, as it we potentially re-allocate numpy arrays twice,
+        # during this padding and the  offset padding before. If that's a bottleneck, we'll optimize.
+        if incoming_num_frames < mix_num_frames:
+            feats_to_add = np.vstack([
+                feats_to_add,
+                self.log_energy_floor * np.ones((mix_num_frames - incoming_num_frames, self.num_features))
+            ])
+
+        # When SNR is requested, find what gain is needed to satisfy the SNR
+        gain = 1.0
+        if snr is not None:
+            # TODO: The SNR stuff here might be tricky. For prototyping assume the following:
+            #  - always dealing with log mel filterbank feats
+            #  - total signal energy can be approximated by summing individual bands energies
+            #    (assumes no spectral leakage...) across time and frequencies
+            #  - achieve the final SNR by determining the signal energy ratio and then scaling
+            #    the signals to satisfy requested SNR
+
+            # Compute the added signal energy before it was padded
+            added_feats_energy = fbank_energy(feats)
```

Comment:
That's an interesting thought. Isn't Parseval's theorem defined in terms of DFT and not STFT? I mean, is there an easy way to reason about the time vs frequency domain energy relationship for a spectrogram as opposed to just DFT?

Reply:
You don't need Parseval's Theorem if you deal with expectations, I think.  Expected energy comes from just summing the energies.

Reply:
> That's an interesting thought. Isn't Parseval's theorem defined in terms of DFT and not STFT? I mean, is there an easy way to reason about the time vs frequency domain energy relationship for a spectrogram as opposed to just DFT?

I think you can group waveform slices in the log, apply jensen inequality and then apply Parseval theorem to each DFT frame independently? I'm not 100% sure. 

> You don't need Parseval's Theorem if you deal with expectations, I think. Expected energy comes from just summing the energies.

Ok. But the SNR range on fbank would be a bit different, no? 
Anyway, it might not be a very important remark from me.